### DISCUSSION TITLE
Netbox intergation with Ansible AWX

### DISCUSSION THREAD

**Discussion by imafukidze** [netbox-community/netbox#11916] [General]
Hello Team!

I am working on my automation solution and have integrated my Netbox instance into Ansible AWX as the inventory. I intend to manage my device through the loopback address. However, I have realized that loopback interface details are not imported into Ansible as (host vars) unless they're associated with physical interfaces on the device.

In the networking world loopback interfaces are not associated with physical interfaces and I was of the opinion that it should be the same in my documentation. 

Does anyone understand why it's designed like that in Netbox and or if the best way to do it?

**Comment by candlerb** [comment]
I think you should set the loopback address as the "primary IP" for each device or VM in Netbox. Is that not visible to Ansible?

**Comment by ryanmerolle** [comment]
@candlerb is correct about the primary IP.

Your other option is to use dns if you have fqdns set for each device the `ansible_host: "{{ inventory_hostname }}.<insert your domain here>"`

You can do this a number of ways:

**1.)** Using `group_vars` specifically in the `all` group
**group_vars/all.yml:**
```yaml
---
ansible_host: "{{ inventory_hostname }}.<insert your domain here>"
```
**2.)** set a `config_context` applied to all platforms you leverage with ansible & enable `config_context` & `flatten_config_context`
**in netbox UI for config context**
```json
{"ansible_host": "{{ inventory_hostname }}.<insert your domain here>"}
```
**netbox_inventory.yml**
```yaml
config_context: true
flatten_config_context: true
```

**3.)** set this in your `netbox_inventory.yml` file with the below:
**netbox_inventory.yml**
```yaml
compose:
  ansible_host: "{{ inventory_hostname }}.<insert your domain here>"
```

**Comment by imafukidze** [comment]
Hello team, thanks for the responses. 

@candlerb, I have tried and its not working. Once I detach the loopback interface from the physical interface I can no longer use it as the device's primary IP. Check the error message on the image below.

<img width="585" alt="NetboxLoopback Error1" src="https://user-images.githubusercontent.com/29435761/223638841-08948337-d26d-4848-9fb8-cffb86f13b9b.png">
<img width="586" alt="NetboxLoopback Error" src="https://user-images.githubusercontent.com/29435761/223636724-6b84deab-f418-4dab-815e-c9679a7872ac.png">

**Reply by candlerb** [reply]
> Once I detach the loopback interface from the physical interface I can no longer use it as the device's primary IP

I think you've misunderstood what a loopback address is.  A loopback interface is not a physical interface, and a loopback address by definition is not associated with any physical interface.

If this really is a loopback address, then what you need to do is:
1. Create a new interface, type "Virtual", name "Loopback0" or whatever, on the device or VM
2. Assign the loopback IP address to that interface
3. Set that IP address as the primary

However, if this is just an IP address on a physical interface, then you can directly set it as the primary IP, and the concept of "loopback" does not come into it at all.

**Reply by imafukidze** [reply]
You saved my day!! 

I followed the steps above and it's now working as I expected.

I know what a loopback interface is but, I didn't realize I had to create it the same way I created physical interfaces on the device and then assign it an IP address. 
My thinking was that devices are not shipped with loopback interfaces but physical interfaces so I created physical interfaces only, hence the error I was facing. I hope you get my point.

Thanks for the help!!

**Reply by candlerb** [reply]
Great! Yes, you need to create any 'virtual' interfaces explicitly in Netbox: this also includes things like VLANs and subinterfaces (Cisco: `int Vlan100` or `int Gi0/0.100` respectively).  That gives you somewhere to hang the IP address.

In the case of a subinterface like `Gi0/0.100`, you can also set the physical interface `Gi0/0` as its "parent" interface.

**Comment by kkthxbye-code** [comment]
> However, I have realized that loopback interface details are not imported into Ansible as (host vars) unless they're associated with physical interfaces on the device.

Unless I'm misunderstanding you, I don't think that's true at all. Create a virtual loopback interface on the device and assign an ip as primary and I believe it should work.
